### ISSUE TITLE
Infinite loop after returning from sleep state

### ISSUE THREAD

**Issue by vaygr** [wsmithril/wslock#3]
Once the box returns from sleep (via `echo "mem" > /sys/power/state`), wslock goes into infinite loop:

```
…
epoll_wait(4, {{EPOLLIN, {u32=3, u64=3}}}, 1, 0) = 1
epoll_wait(4, {{EPOLLIN, {u32=3, u64=3}}}, 1, 0) = 1
epoll_wait(4, {{EPOLLIN, {u32=3, u64=3}}}, 1, 0) = 1
epoll_wait(4, {{EPOLLIN, {u32=3, u64=3}}}, 1, 0) = 1
epoll_wait(4, {{EPOLLIN, {u32=3, u64=3}}}, 1, 0) = 1
epoll_wait(4, {{EPOLLIN, {u32=3, u64=3}}}, 1, 0) = 1
epoll_wait(4, {{EPOLLIN, {u32=3, u64=3}}}, 1, 0) = 1
epoll_wait(4, {{EPOLLIN, {u32=3, u64=3}}}, 1, 0) = 1
epoll_wait(4, {{EPOLLIN, {u32=3, u64=3}}}, 1, 0) = 1
…
```

**Comment by vaygr** [comment]
After looking into this a bit closer it seems it's different from the issue fixed by 1dd0e71.

Looks like we end up infinitely [at `EINTR`](https://github.com/wsmithril/wslock/blob/1a472af756924245696621f54129fc0807d03cf2/wslock.c#L450). But not only when the loop is interrupted on suspend, but the same kind of behavior can be seen when you simply attach to the process with e.g. `strace`.

To me it looks like in this specific case `epoll_wait()` doesn't really reset `errno` on successful run (nor that it should) so we keep relying on the old value from when PC went from sleep / we attached to the process, thus infinite loop.

Here's a quote from `errno(3)`:

```
       For  some system calls and library functions (e.g., getpriority(2)), -1
       is a valid return on success.  In such cases, a successful  return  can
       be  distinguished  from an error return by setting errno to zero before
       the call, and then, if the call returns a status that indicates that an
       error may have occurred, checking to see if errno has a nonzero value.
```

I looked at how other projects handle it. In Varnish for example they don't have that kind of reset when `epoll_wait()` is in the loop, they have a warning though with a condition (snippet from `bin/varnishd/waiter/cache_waiter_epoll.c`):

```c
		do {
			/* Due to a linux kernel bug, epoll_wait can
			   return EINTR when the process is subjected to
			   ptrace or waking from OS suspend. */
			n = epoll_wait(vwe->epfd, ev, NEEV, i);
		} while (n < 0 && errno == EINTR);
```

In original Python interpreter they do reset `errno` to zero (snippet from `Modules/selectmodule.c`):

```c
    do {
        Py_BEGIN_ALLOW_THREADS
        errno = 0;
        nfds = epoll_wait(self->epfd, evs, maxevents, (int)ms);
        Py_END_ALLOW_THREADS

        if (errno != EINTR)
            break;

        /* poll() was interrupted by a signal */
        if (PyErr_CheckSignals())
            goto error;

        if (timeout >= 0) {
            timeout = deadline - _PyTime_GetMonotonicClock();
            if (timeout < 0) {
                nfds = 0;
                break;
            }
            ms = _PyTime_AsMilliseconds(timeout, _PyTime_ROUND_CEILING);
            /* retry epoll_wait() with the recomputed timeout */
        }
    } while(1);
```

I couldn't debug it any further but my PR mentioned above fixes the issue perfectly. Tested for both cases.

**Comment by vaygr** [comment]
To add, i3lock doesn't have this problem simply because libev handles the issue for them (snippet from `ev_epoll.c`):

```c
  EV_RELEASE_CB;
  eventcnt = epoll_wait (backend_fd, epoll_events, epoll_eventmax, timeout * 1e3);
  EV_ACQUIRE_CB;

  if (expect_false (eventcnt < 0))
    {
      if (errno != EINTR)
        ev_syserr ("(libev) epoll_wait");

      return;
    }
```
